### DISCUSSION TITLE
PKI Implementations

### DISCUSSION THREAD

**Discussion by JavaJoeS** [eclipse-m2e/m2e-core#1538] [General]
Why are there no hooks or supporting packages for any PKI, PKCS12, and more specifically PKCS11?
Can I do anything to assist in developing hooks or service implementations to support PKI?
Eclipse has moved towards https and now most, if not all connections are https.  Makes sense to add
the capability for support.  I have a customized set of packages Im willing to share!

**Comment by JavaJoeS** [comment]
What code would I modify if I were implementing?  To start 4 packages
need impl.
1. org.eclipse.m2e.maven.runtime
2. org.eclipse.m2e.launching
3. org.eclipse.m2e.core 
4. org.eclipse.m2e.core.ui

**Reply by laeubi** [reply]
> What code would I modify if I were implementing? 

The very first thing would be to precisely describe the problem and the proposed solution so that others can understand what you are trying to archive.

**Reply by JavaJoeS** [reply]
The problem is NO support for any PKI and the solution to that would be an implementation.  Probably many to choose from.
PKCS12 would be pretty simple to start, and then PKCS11 has OPEN source version implementation that could be integrated.
I surely do not know the M2E code as the authors, or where it all is integrated, besides eclipse.  You guys are genius and probably 
have already thought about this! lol!

**Reply by JavaJoeS** [reply]
I have done a quick and dirty impl with PKCS12 and using CSPid PKCS11 package.  However, my impl is specific.  Thinking
out of the box, a more generic approach is needed so that whatever a company uses can easily be added via hooks or SVC
impls...

**Reply by HannesWell** [reply]
> The problem is NO support for any PKI and the solution to that would be an implementation.

But what do you want to do you want to do the Public Key Infrastructure? E.g. access a protected repository?

**Reply by JavaJoeS** [reply]
My idea is that add generic support for PKI with all the standard
java implements, i.e. SSLContext and all the java JAVAX properies
if set at all.  Default is NO PKI unless a SVC or SVC Extension is
installed.   My initial implementation used a GENERIC pkcs12
impl and a Custom CSPid PKCS11 implementation.  I added
packages to runtime, implementation to the Launcher and package
extensions in core and core.ui.  WOrkwed Like a champ.
It allows you to use a embeded Maven (modified slightly for PKCS11)
and hooked to the default MAVEN PKCS12, already in maven.

**Reply by JavaJoeS** [reply]
Yes, allows eclipse and maven to access https secured sites, customer uses their own
keystore and truststore.

**Reply by laeubi** [reply]
You should be able to set default key/trustores in the eclipse.ini

**Reply by JavaJoeS** [reply]
I created a custom ui wizard in eclipse to allow a user to add either a pkcs11 or pkcs12 Keystore/pin.  I updated 
eclipse to be CSPid PKCS11 enabled, the spec allows for any app to automatically connect up if the cspid manager
for pkcs11 is already on.  Then there is no need for either adding anything to the ini or popping up a wizard.

**Reply by laeubi** [reply]
If you open the preferences and search for SSH you already see some settings for `SSH (Public Key)`, I think something like this for SSL in [Eclipse Platform](https://github.com/eclipse-platform/eclipse.platform) first then m2e can reuse that.

**Reply by JavaJoeS** [reply]
I had to include an instance of my security app inside of the m2e.launcher package, specifically update required
to the CLI class.  There was no properties that were handed off to maven, nore was the pin available.

**Reply by laeubi** [reply]
As said first platform should support PKI then we can discuss on how to integrate it with maven/m2e ...

**Reply by mickaelistria** [reply]
Yes, such a page should be contributed into Platform, and appear under General > Network.

**Reply by JavaJoeS** [reply]
Can you help me to achieve the process of contributing and implementation!  Thank you
very much for looking.

**Reply by mickaelistria** [reply]
Please see https://github.com/eclipse-platform/.github/blob/main/CONTRIBUTING.md and bring further questions to Platform.

**Comment by JavaJoeS** [comment]
![Screenshot-PKI-2023-09-12 10-59-08](https://github.com/eclipse-m2e/m2e-core/assets/21100871/d2c35f7b-cb91-4278-b6d4-1d779ed43bc3)

**Reply by JavaJoeS** [reply]
I have the screenshot above will load in eclipse, unless a CSPid manager has already been loaded.
CSPid enabled applications will automatically be PKI enabled.  I CSPid enabled eclipse by adding 
referenced packages into it.

**Reply by JavaJoeS** [reply]
@HannesWell 
Check out my screenshot!

**Comment by JavaJoeS** [comment]
Can you help me get started?Sent from my iPadOn Sep 13, 2023, at 3:02 AM, Mickael Istria ***@***.***> wrote:﻿
Please see https://github.com/eclipse-platform/.github/blob/main/CONTRIBUTING.md and bring further questions to Platform.

**Reply by mickaelistria** [reply]
Feel free to ask any technical question and people may help.
